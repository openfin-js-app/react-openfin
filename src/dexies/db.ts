import Dexie from 'dexie';

import { IConfigDexie } from '../reduxs';

class OpenfinReactDatabase extends Dexie{
    configs:Dexie.Table<IConfigDexie,number>;

    constructor(databaseName){
        super(databaseName);
        this.version(1).stores({
            configs:'++id,tabName,fieldName'
        })
        this.version(2).stores({
            configs:'++id,tabName,fieldName,version'
        })
    }
}

const db = new OpenfinReactDatabase('openfinReactDb');

db.open().catch((err)=>{
    console.error('Default openfinReactDb db failed',err);
});

export default db;