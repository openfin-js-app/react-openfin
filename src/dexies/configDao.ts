import db from './db';

import { IConfigDexie } from '../reduxs';


export const CONFIG_VERSION:number = 2;

export async function findAll():Promise<IConfigDexie[]>{
    return  db.table('configs').toArray();
}

export async function findAllOfCurrentVersion():Promise<IConfigDexie[]> {
    return db.table('configs')
        .where({version:CONFIG_VERSION})
        .toArray();
}

export async function findOneField(tabName:string,fieldName:string):Promise<IConfigDexie> {
    const configs = db.table('configs');
    return await configs.where({
        tabName,fieldName, version:CONFIG_VERSION,
    }).first();
}

export async function findOneFieldVal(tabName:string,fieldName:string):Promise<any> {
    const oneItem:IConfigDexie = await findOneField(tabName,fieldName);
    return oneItem?oneItem.value:null;
}

export async function saveOrUpdateOneByTabNameFieldName(tabName:string, fieldName:string, value:any):Promise<IConfigDexie> {
    let one:IConfigDexie = null;
    const configs = db.table('configs');
    await db.transaction('rw',configs,async()=>{
        const founds:IConfigDexie[] = await configs
            .where({tabName,fieldName,version:CONFIG_VERSION})
            .toArray();
        // console.log('configDao::saveOrUpdateOneByTabNameFieldName',founds.length,founds,tabName,fieldName,value);
        if (founds.length === 0){
            one = {tabName,fieldName,value,version:CONFIG_VERSION};
            const id = await configs.put(one);
            one = await configs.get(id);
        }else if (founds.length === 1){
            one = founds[0];
            await configs.update(one.id,{value});
            one = await configs.get(one.id);
        }else{
            await configs.bulkDelete(founds.map(one => one.id));
            one = {tabName,fieldName,value,version:CONFIG_VERSION};
            const id = await configs.put(one);
            one = await configs.get(id);
        }
    });
    return one;
}