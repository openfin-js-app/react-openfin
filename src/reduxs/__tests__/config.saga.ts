import { testSaga,expectSaga } from 'redux-saga-test-plan';
import * as matchers from 'redux-saga-test-plan/matchers'
import {call, delay, put, select, take, takeEvery, takeLatest,} from 'redux-saga/effects';

import {
    CONFIG_UPDATE_NEW_WINDOW_POSITION,
    configDoUpdateOneFieldInDexie,
    configUpdateNewWindowPositionAddDelta,
    configUpdateNewWindowPositionResetLeft,
    configUpdateNewWindowPositionResetTop,
} from '..';
import {System, Window} from "redux-openfin";

import {
    // selectors
    getNewWindowTop,
    getNewWindowLeft,
    getNewWindowWidth,
    getNewWindowHeight,
    // sub sagas
    handleConfigUpdateNewWindowPosition,
    handleConfigLoadFromDexie,
    handleConfigUpdateOneField,
    handleConfigUpdateOneFieldInDexie
} from '../sagas/config';

import configSaga from '../sagas/config';
import {
    CONFIG_DO_UPDATE_ONE_FIELD_IN_DEXIE, CONFIG_LOAD_FROM_DEXIE, CONFIG_UPDATE_ONE_FIELD,
    configDoUpdateOneField
} from "..";
import {findAllOfCurrentVersion,saveOrUpdateOneByTabNameFieldName} from "../../dexies/configDao";

describe('Config saga',()=>{

    describe('handleConfigLoadFromDexie saga',()=>{
        it('it basically works',()=>{

            testSaga(handleConfigLoadFromDexie)
                .next()
                .call(findAllOfCurrentVersion)
                .next([{tabName:'tabName',fieldName:'fieldName',value:'value'}])
                .put(configDoUpdateOneField({
                    tabName:'tabName',fieldName:'fieldName',value:'value'
                }))
                .next()
                .isDone();
        })
    });

    describe('handleConfigUpdateOneField saga',()=>{
        it('it basically works',()=>{

            testSaga(handleConfigUpdateOneField,{payload:{name:'tabName.fieldName',value:'value'}})
                .next()
                // @ts-ignore
                .putResolve(configDoUpdateOneField({
                    tabName:'tabName',fieldName:'fieldName',value:'value'
                }))
                .next()
                .put(configDoUpdateOneFieldInDexie({
                    tabName:'tabName',fieldName:'fieldName',value:'value'
                }))
                .next()
                .isDone();
        })
    });

    describe('handleConfigUpdateOneFieldInDexie saga',()=>{
        it('it basically works',()=>{
            expectSaga(handleConfigUpdateOneFieldInDexie,{payload:{tabName:'tabName',fieldName:'fieldName',value:'value'}})
                .call(saveOrUpdateOneByTabNameFieldName,'tabName','fieldName','value')
                .delay(800)
                .run();
        })
    });

    describe('handleConfigUpdateNewWindowPosition saga',()=>{
        it('add delta to new win pos',()=>{
            const newWinWidth = 400;
            const newWinHeight = 300;
            const newWinTop = 10;
            const newWinLeft = 10;
            const virtualScreen = {
                top:0, left:0, right:800, bottom:600,
            };
            testSaga(handleConfigUpdateNewWindowPosition)
                .next()
                .select(getNewWindowWidth)
                .next(newWinWidth)
                .select(getNewWindowHeight)
                .next(newWinHeight)
                .select(getNewWindowTop)
                .next(newWinTop)
                .select(getNewWindowLeft)
                .next(newWinLeft)
                .call(System.asyncs.getMonitorInfo,System.actions.getMonitorInfo({}))
                .next({payload:{ virtualScreen }})
                // @ts-ignore
                .putResolve(configUpdateNewWindowPositionAddDelta())
                .next()
                .isDone();
        });
        it('reset left of the new win pos',()=>{
            const newWinWidth = 400;
            const newWinHeight = 300;
            const newWinTop = 10;
            const newWinLeft = 410;
            const virtualScreen = {
                top:0, left:0, right:800, bottom:600,
            };
            testSaga(handleConfigUpdateNewWindowPosition)
                .next()
                .select(getNewWindowWidth)
                .next(newWinWidth)
                .select(getNewWindowHeight)
                .next(newWinHeight)
                .select(getNewWindowTop)
                .next(newWinTop)
                .select(getNewWindowLeft)
                .next(newWinLeft)
                .call(System.asyncs.getMonitorInfo,System.actions.getMonitorInfo({}))
                .next({payload:{ virtualScreen }})
                // @ts-ignore
                .putResolve(configUpdateNewWindowPositionResetLeft())
                .next()
                .isDone();
        });
        it('reset top of the new win pos',()=>{
            const newWinWidth = 400;
            const newWinHeight = 300;
            const newWinTop = 310;
            const newWinLeft = 10;
            const virtualScreen = {
                top:0, left:0, right:800, bottom:600,
            };
            testSaga(handleConfigUpdateNewWindowPosition)
                .next()
                .select(getNewWindowWidth)
                .next(newWinWidth)
                .select(getNewWindowHeight)
                .next(newWinHeight)
                .select(getNewWindowTop)
                .next(newWinTop)
                .select(getNewWindowLeft)
                .next(newWinLeft)
                .call(System.asyncs.getMonitorInfo,System.actions.getMonitorInfo({}))
                .next({payload:{ virtualScreen }})
                // @ts-ignore
                .putResolve(configUpdateNewWindowPositionResetTop())
                .next()
                .isDone();
        });
    });

    it('default function register all event',()=>{
        testSaga(configSaga)
            .next()
            .takeEvery(CONFIG_LOAD_FROM_DEXIE, handleConfigLoadFromDexie)
            .next()
            .takeEvery(CONFIG_UPDATE_ONE_FIELD, handleConfigUpdateOneField)
            .next()
            .takeLatest(CONFIG_DO_UPDATE_ONE_FIELD_IN_DEXIE, handleConfigUpdateOneFieldInDexie)
            .next()
            .takeLatest(CONFIG_UPDATE_NEW_WINDOW_POSITION,handleConfigUpdateNewWindowPosition)
            .next()
            .isDone();
    })
});
