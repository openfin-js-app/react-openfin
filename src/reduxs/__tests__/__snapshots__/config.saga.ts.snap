// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Config saga handleConfigSelectOneField dynamic selector function 1`] = `"value"`;

exports[`Config saga handleConfigSelectOneField dynamic selector function 2`] = `null`;

exports[`Config saga handleConfigSelectOneField dynamic selector function 3`] = `null`;
